rfcs/permission-improvements-v3.md

```diff
@@ -0,0 +1,330 @@
+be specified in a request.
+
+For example, if a request's authz resolves to the following roles:
+
+```yaml
+x-hasura-role: reader, editor
+```
+
+graphql-engine would combine the permissions of both reader and editor when
+processing the request.
+
+### 2.1 Challenges:
+
+Composing permissions this way seems quite natural. Let's see how this gets
+complicated.
+
+#### 2.1.1 No overlapping permissions
+If there are no overlapping permissions for a table, the merge semantics
+are straightforward.
+
+#### 2.1.2 Overlapping permissions
+
+i.e. permission on a table is defined for two different roles:
+
+1. For selects:
+
+    1. If the column sets are the same, this is easy. The 'filter' becomes
+   `filter(r1)` OR `filter(r2)`.
+
+    1. If the column sets are not the same, for example:
+
+       ```yaml
+       - role: public
+         table: users
+         columns: id, username
+         filter: {}
+
+       - role: private
+         table: users
+         columns: id, username, email
+         filter:
+           id: x-hasura-user-id
+       ```
+
+       What should the semantics of this query be?
+
+       ```graphql
+       query {
+           users {
+               id
+               username
+               email
+           }
+       }
+       ```
+
+       If we were to union the columns and the rows, the users get to see
+       `email` column for all users and not just theirs.
+
+       > The expected behaviour however is that `email` is returned as `null`
+       for all other users.
+
+       This is fairly challenging to implement, v2 uses SQL `case` expressions to
+       suport this and this heavily relies on the database features. It may not be
+       possible to have these correct semantics for all databases.
+
+1. For mutations, if the presets are different, we do not know how which one to
+   use.
+
+### 2.2 Merge/Conflict strategies
+
+In v2, permission composition happens at the metadata layer, any invalid
+combination of roles would result in a failed metadata update.
+
+Once we start allowing multiple roles per request, we need to figure out the
+strategy for handling these overlapping permissions:
+
+1. Fail the request:
+
+   If the operation being invoked has more than one permission, we fail the
+   request. This is quite straightforward but may not be the desired behaviour.
+
+2. Provide merge semantics:
+
+   In the metadata, instead of composing a fixed set of permissions using
+   inherited roles, we could for example, allow specifying the permission
+   capabilities. For example,
+
+    ```yaml
+    conflict_rules:
+    # admin insert permissions should be preferred over user role's insert
+    # permissions on all tables
+    - admin.insert > user.insert
+
+    # manager's update permissions should be preferred over user role's select
+    # update permissions on employee_details table
+    - manager.update.employee_details > user.update.employee_details
```

Comment:
I think this shouldn't be part of the metadata (otherwise we will run into the problems we had with inherited roles in V2) and rather it should be part of the GraphQL request, akin to SQL's `on_conflict`. 

I am not sure how this information can be provided on a per-request basis, but one idea I can think of which may work is to accept a header that accepts a JSON object with the above mentioned rules.